src/controllers/question.js

```diff
@@ -1,11 +1,120 @@
+// const mongoose = require('mongoose');
+const Question = require('../models/question');
+const Student = require('../models/student');
+
+// function filterQuestions(query) {
+//   return async (_, res) => {
+//     try {
+//       // Find and sort by latest
+//       const questions = await Question.find(query); // .sort({ createdAt: -1 });
+//       if (questions) res.status(201).json(questions);
+//       else res.status(200).json({ messge: 'No questions found' });
+//     } catch (err) {
+//       res.status(422).json({ message: err.message });
+//     }
+//   };
+// }
+
+const filterQuestions = async (res, query) => {
+  try {
+    // Find and sort by latest
+    const questions = await Question.find(query); // .sort({ createdAt: -1 });
+    if (questions.length <= 0) {
+      res.status(200).json({ message: 'No questions found' });
+    } else res.status(200).json(questions);
+  } catch (err) {
+    res.status(422).json({ message: err.message });
+  }
+};
 module.exports = {
-  getAllQuestions: async () => {},
-  getOneQuestion: async () => {},
-  addNewQuestion: async () => {},
+  getAllQuestions: async (req, res) => {
+    /* test if it's registered user then get all the questions
+      related to the interested subjects of the user
+     */
+    const query = {};
+    if (req.user) {
+      // find user's subjects
+      // const userId = req.auth.sub;
+      const subjects = await Student.findById(req.user.id).select({
+        subjects: 1,
+      });
+      // if user has no subject then get all questions ordered by date
+      if (!subjects) filterQuestions(res, query);
+      else {
+        query.subjetc = subjects;
+        filterQuestions(res, query);
+      }
+    }
+    // if not registered user get all questions ordered by date (recent questions)
+    else {
+      // console.log('Heeeeellllo');
+      filterQuestions(res, query);
+    }
+  },
+
+  getOneQuestion: async (req, res) => {
+    const { id } = req.params;
+    try {
+      const question = await Question.findById(id);
+      if (question.length <= 0)
+        res
+          .status(200)
+          .json({ message: `The question you are looking for not found` });
+      else res.status(200).json(question);
+    } catch (err) {
+      res.status(422).json({ message: err.message });
+    }
+    // }
+  },
+  addNewQuestion: async (req, res) => {
+    /*
+    To add question the user needs to be registered(onlyAtuhenticated middleware check that)
+    All necessary validations are handled by quesion model
+    so we can save new questions directly
+    */
```

Comment:
Yes, that's right but if we didn't enter all the fields, some of them would not be generated in the document in the database, so I initialized them with empty values to ensure their existence. such as the "subjects" array. Because it's not required, it will pass the validation. I thought it would be much better to initialize all the fields of the document.